Update contributing guidelines

kahmali opened the pull request:
- Add more detailed information to guidelines (using the guidelines from
  Restivus as a starting point)
- Organize information into sections
- Enforce the 100 character line limit in document (to avoid hypocrisy)

stubailo commented on CONTRIBUTING.md:
```diff
@@ -1,11 +1,108 @@
+Please adhere to the [Meteor Style Guide](https://github.com/meteor/meteor/wiki/Meteor-Style-Guide) 
+for all conventions not specified here:
+
+1. 100 character line limit for all code, comments, and documentation files
```
I'd like to stick to 80 chars, please.

kahmali replied:
I will reluctantly accept this. :) My personal view on it is that 80 chars is an [outdated standard](http://programmers.stackexchange.com/questions/148677/why-is-80-characters-the-standard-limit-for-code-width). Where is code being read anymore where an 80 char line limit is imposed (actually, see edit below)? There is nothing about 80 chars that is more readable, and even big companies like Google have seen the light and adopted the 100 char line limit in major projects. From their [Android Coding Guidelines](https://source.android.com/source/code-style.html): 
```text
Each line of text in your code should be at most 100 characters long.

There has been lots of discussion about this rule and the decision remains that 100 characters is
the maximum.

Exception: if a comment line contains an example command or a literal URL longer than 100
characters, that line may be longer than 100 characters for ease of cut and paste.
```

My argument against the 80 char line limit is that I sometimes find myself unnaturally breaking up a line of code, which makes it less readable. My goal is always code readability, to make maintaining the codebase easier. 

If we stick with the 80 char limit for code and comments, could we consider using 100 in the markdown files? If we don't, we may want to use named links instead of inlining them, since URLs are typically long and sometimes span more than 80 chars. And at the very least we should make the same exception as Google for URLs longer than our line limit. I struggled with the 80 char limit for a while in my personal projects, and have since adopted the 100 everywhere. Alright, that's probably enough debate about line limit for one day. :D

**EDIT:** I stand corrected. Apparently, GitHub appears to be cutting off lines of `code` at 80 chars here in this thread (at least in chrome for me right now). Definitely an argument in favor of the 80 char limit.

stubailo commented on CONTRIBUTING.md:
```diff
@@ -1,11 +1,108 @@
+   1. Ensure all automated tests are passing by running `meteor test-packages ./` from the root
+      directory of the project and viewing the Tinytest output at `http://localhost:3000`
+   1. Update [README](https://github.com/stubailo/meteor-rest/blob/devel/packages/rest/README.md) 
+      and [change log](https://github.com/stubailo/meteor-rest/blob/devel/packages/rest/README.md#change-log) 
```
Apparently, GitHub has a special file name for change logs. Do people actually use that?

kahmali replied:
Actually, [I do on Restivus](https://github.com/kahmali/meteor-restivus/blob/devel/CHANGELOG.md). I haven't come across any GitHub integration for it yet, but I did read a [great article](http://keepachangelog.com/) that imposed a solid standard that I've followed since beginning Restivus, and have really come to love. I was planning to break ours up into separate files in a separate PR soon, if you're okay with it. We can link to the change logs from the READMEs.

stubailo commented:
This is awesome! Some comments inline.

aldeed commented on CONTRIBUTING.md:
```diff
@@ -1,11 +1,108 @@
+1. Before submitting a pull request:
+   1. Rebase to the latest version of `devel`
+   1. Add automated tests for any new features 
+   1. Ensure all automated tests are passing by running `meteor test-packages ./` from the root
```
In this project, running tests should be done with `./run-tests.sh`

aldeed commented:
I like it. I made one small comment.

I prefer no hard line length limits at all, but because it's a never-ending debate and because this will ideally become THE Meteor rest package, following MDG style guide strictly makes sense to me.

kahmali commented:
@aldeed That makes perfect sense. I went ahead and removed the reference to the 100 character line limit, and updated the file to use an 80 char limit. I also fixed the test command, per your suggestion.

@stubailo I made a few other small changes. I kept the commits separate for now so the minor changes between the two would be easier to see. The [rich diff](https://github.com/stubailo/meteor-rest/commit/f0cbe219b8f8589068aa0f7541a0101689130c40?short_path=6a33714#diff-6a3371457528722a734f3c51d9238c13) is really handy for spotting those changes (since the line limit changes won't get in the way).

aldeed commented:
LGTM

stubailo commented:
I just noticed this part:

> Use single-quoted strings (e.g., 'Hello world!' instead of "Hello world!")

I use exclusively double quotes everywhere - are you going to go through the existing code and change all of the quotes?

kahmali commented:
@stubailo I just noticed that you've tried to adhere to double-quoted strings in most places. Meteor autogenerates the package file with single-quoted strings, and I did notice some single quoted strings in a few other places (mostly in calls to `Npm.require()`). My only goal is consistency. I'm willing to convert all strings to a single style, whichever we choose. By habit, I always find myself using single-quotes (less "bulky" I guess), but I originally come from Java and am certainly not opposed to double-quoted strings.

aldeed commented:
I'm OK with either, but I've been favoring single quotes lately.

stubailo commented:
OK sure I think most people like single quotes more, I'll try to stay with those.

kahmali commented:
There's an extra commit in here (and a merge where we could have rebased), but it's too late to change it unless we're willing to rewrite history. I'm sure it'll be fine. Just my insane git OCD nagging me.

stubailo commented:
Oh interesting - at Meteor we usually make the merges explicit in the case
of actually merging new features, so there is a record. But I can go either
way.
On Wed, May 27, 2015 at 2:22 AM Kahmali Rose <notifications@github.com>
wrote:

> There's an extra commit in here (and a merge where we could have rebased),
> but it's too late to change it unless we're willing to rewrite history. I'm
> sure it'll be fine. Just my insane git OCD nagging me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stubailo/meteor-rest/pull/27#issuecomment-105772438>.
>
